File: src/main/cpp/plazma/lib/data/xml/xmllib.cpp
#include <string>

#include "xmllib.h"
#include "xmltokenizer.h"
#include "xmlparser.h"
#include "xmlwriter.h"
#include "Xml2JsonConverter.h"
#include "Xml2YamlConverter.h"

#include "plazma/lib/io/iolib.h"
#include "plazma/lib/text/textlib.h"

using namespace ext;

namespace xmllib {

    // tokenize

    StringList* tokenizeXmlFromText(const std::string& input) {
        return xmltokenizer::tokenizeXmlFromArray(input.c_str(), input.length());
    }

    StringList* tokenizeXmlFromArray(const char* input) {
        return xmltokenizer::tokenizeXmlFromArray(input);
    }

    StringList* tokenizeXmlFromArray(const char* input, int size) {
        return xmltokenizer::tokenizeXmlFromArray(input, size);
    }

    StringList* tokenizeXmlFromFile(const std::string& fileName) {
        char *text = iolib::readChars(fileName);
        return tokenizeXmlFromArray(text);
    }

    // parse
    
    node::Node* parseXmlFromTokens(StringList* tokens) {
        return xmlparser::parseXmlFromTokens(tokens);
    }

    node::Node* parseXmlFromTokens(xml::XmlReaderConfig* config, StringList* tokens) {
        return xmlparser::parseXmlFromTokens(config, tokens);
    }

    // read

    node::Node* readXmlFromText(const std::string& input) {
        StringList* tokens = xmltokenizer::tokenizeXmlFromText(input);
        return parseXmlFromTokens(tokens);
    }

    node::Node* readXmlFromText(xml::XmlReaderConfig* config, const std::string& input) {
        StringList* tokens = xmltokenizer::tokenizeXmlFromText(input);
        return parseXmlFromTokens(config, tokens);
    }

    //

    node::Node* readXmlFromArray(const char* input) {
        StringList* tokens = xmltokenizer::tokenizeXmlFromArray(input);
        return parseXmlFromTokens(tokens);
    }

    node::Node* readXmlFromArray(xml::XmlReaderConfig* config, const char* input) {
        StringList* tokens = xmltokenizer::tokenizeXmlFromArray(input);
        return parseXmlFromTokens(config, tokens);
    }

    //

    node::Node* readXmlFromFile(const std::string& fileName) {
        char *text = iolib::readChars(fileName);
        return readXmlFromArray(text);;
    }

    node::Node* readXmlFromFile(xml::XmlReaderConfig* config, const std::string& fileName) {
        char *text = iolib::readChars(fileName);
        return readXmlFromArray(config, text);;
    }

    // write: text

    std::string writeXmlToText(node::Node* root) {
        return xmlwriter::writeXmlToText(root);
    }

    std::string writeXmlToText(xml::XmlWriterConfig* config, node::Node* root) {
        return xmlwriter::writeXmlToText(config, root);
    }

    // write: console

    void writeXmlToConsole(node::Node* root) {
        xmlwriter::writeXmlToConsole(root);
    }

    void writeXmlToConsole(xml::XmlWriterConfig* config, node::Node* root) {
        xmlwriter::writeXmlToConsole(config, root);
    }

    // write: file

    void writeXmlToFile(const std::string& fileName, node::Node* root) {
        // TODO: Use xmlwriter::writeXmlToFile(fileName, config, root);
        std::string text = writeXmlToText(root);
        iolib::writeText(fileName, text);
    }

    void writeXmlToFile(const std::string& fileName, xml::XmlWriterConfig* config, node::Node* root) {
        // TODO: Use xmlwriter::writeXmlToFile(fileName, config, root);
        std::string text = writeXmlToText(config, root);
        iolib::writeText(fileName, text);
    }

    // convert

    node::Node* convertXmlToJson(node::Node* node) {
        xml::Xml2JsonConverter* converter = new xml::Xml2JsonConverter();
        node::Node* result = converter->convert(node);
        if (converter != nullptr) {
            delete converter;
        }        
        return result;
    }

    node::Node* convertXmlToYaml(node::Node* node) {
        xml::Xml2YamlConverter* converter = new xml::Xml2YamlConverter();
        node::Node* result = converter->convert(node);
        if (converter != nullptr) {
            delete converter;
        }        
        return result;
    }

    //

    std::string getTokensText(StringList* tokens) {
        return xmltokenizer::getTokensText(tokens);
    }



}

File: src/main/java/plazma/lib/data/xml/XmlLib.java
/*
 * Copyright (C) 2012-2023 Oleh Hapon ohapon@users.sourceforge.net
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Oleh Hapon
 * Kyiv, UKRAINE
 * ohapon@users.sourceforge.net
 */

package plazma.lib.data.xml;

import java.io.IOException;

import plazma.lib.data.node.Node;

public class XmlLib {

    // tokenize

    public static String[] tokenizeXmlFromText(String input) {
        return new XmlTokenizer().tokenizeXmlFromText(input);
    }

    public static String[] tokenizeXmlFromArray(char[] input) {
        return new XmlTokenizer().tokenizeXmlFromArray(input);
    }

    // parse

    public static Node parseXmlFromTokens(String[] tokens) {
        return new XmlParser().parseXmlFromTokens(tokens);
    }

    public static Node parseXmlFromTokens(XmlReaderConfig config, String[] tokens) {
        return new XmlParser().parseXmlFromTokens(config, tokens);
    }

    // read

    public static Node readXmlFromText(String input) {
        return new XmlReader().readXmlFromText(input);
    }

    public static Node readXmlFromText(XmlReaderConfig config, String input) {
        return new XmlReader().readXmlFromText(config, input);
    }

    //

    public static Node readXmlFromArray(char[] input) {
        return new XmlReader().readXmlFromArray(input);
    }

    public static Node readXmlFromArray(XmlReaderConfig config, char[] input) {
        return new XmlReader().readXmlFromArray(config, input);
    }

    //

    public static Node readXmlFromFile(String fileName) throws IOException {
        return new XmlReader().readXmlFromFile(fileName);
    }

    public static Node readXmlFromFile(XmlReaderConfig config, String fileName) throws IOException {
        return new XmlReader().readXmlFromFile(config, fileName);
    }

    // write

    public static String writeXmlToText(Node root) {
        return new XmlWriter().writeXmlToText(root);
    }

    public static String writeXmlToText(XmlWriterConfig config, Node root) {
        return new XmlWriter().writeXmlToText(config, root);
    }
    
    //
    
    public static void writeXmlToFile(String fileName, Node root) throws IOException {
        new XmlWriter().writeXmlToFile(fileName, root);
    }

    public static  void writeXmlToFile(String fileName, XmlWriterConfig config, Node root) throws IOException {
        new XmlWriter().writeXmlToFile(fileName, config, root);
    }

    //

    public static void writeXmlToConsole(Node root) throws IOException {
        new XmlWriter().writeXmlToConsole(root);
    }

    public static void writeXmlToConsole(XmlWriterConfig config, Node root) throws IOException {
        new XmlWriter().writeXmlToConsole(config, root);
    }

    // convert

    public static Node convertXmlToJson(Node node) {
        Xml2JsonConverter converter = new Xml2JsonConverter();
        Node result = converter.convert(node);
        //if (converter != null) {
        //    converter.destroy();
        //}        
        return result;
    }

    public static Node convertXmlToYaml(Node node) {
        Xml2YamlConverter converter = new Xml2YamlConverter();
        Node result = converter.convert(node);
        //if (converter != null) {
        //    converter.destroy();
        //}        
        return result;
    }

}
